sdk/client.py

```diff
@@ -146,21 +41,66 @@ def __init__(
         password: Optional[str] = None,
         token_provider: Optional[TokenProvider] = None,
         endpoints: Optional[Endpoints] = None,
+        auth0: Optional[Auth0Conf] = None,
     ):
-        if not (username and password) and not token_provider:
-            raise Exception(
-                "At least a username/password combination or"
-                "TokenProvider object should be provided."
+        if not username and not token_provider:
+            raise ValueError(
+                "At least a username or TokenProvider object should be provided."
             )
-        self.endpoints = endpoints or Endpoints()
-        self.group_id = group_id
+        if token_provider is not None:
+            self._check_token_provider(token_provider)
 
-        if username and password:
-            token_provider = UsernamePasswordTokenProvider(
-                username, password, f"{self.endpoints.account}/api/v1/auth/login"
-            )
+        self.endpoints = self._make_endpoints(endpoints)
+
+        if username:
+            auth0 = self._make_auth0(auth0)
+            token_provider = self._credential_login(username, password, auth0)
 
         self.authenticator = HTTPBearerAuthenticator(token_provider)
+        self.group_id = group_id
+
+    @staticmethod
+    def _make_endpoints(endpoints: Optional[Endpoints]) -> Endpoints:
+        if endpoints is None:
+            return Endpoints()
+
+        if not isinstance(endpoints, Endpoints):
+            raise TypeError(f"endpoints must be a {Endpoints.__name__} instance")
+
+        return endpoints
+
+    @staticmethod
+    def _make_auth0(auth0: Optional[Auth0Conf]) -> Auth0Conf:
+        if auth0 is None:
+            return Auth0Conf()
+
+        if not isinstance(auth0, Auth0Conf):
+            raise TypeError(f"auth0 parameter must be a {Auth0Conf.__name__} instance")
+
+        return auth0
+
+    @staticmethod
+    def _check_token_provider(token_provider: TokenProvider) -> None:
+        try:
+            # The type ignore is because I wouldn't know how to fix the type problem
+            # but the code should be correct, and is tested
+            issubclass(token_provider, TokenProvider)  # type: ignore
+        except TypeError:
+            raise TypeError("token_provider must be a TokenProvider subclass")
+
+    def _credential_login(
+        self, username: str, password: Optional[str], auth0: Auth0Conf
+    ) -> TokenProvider:
+        if not password:
+            password = getpass("Enter your password:")
+
+        #   We want to allow an empty string at first, but then
```

Comment:
nit: "The prompted password should not be empty"

Reply:
done